pkg/auth/providers/azuread/azuread_provider.go

```diff
@@ -0,0 +1,291 @@
+	"github.com/pkg/errors"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/rancher/types/user"
+
+	"net/http"
+
+	"strings"
+
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/types/apis/management.cattle.io/v3public"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	Name = "azuread"
+)
+
+type azureADProvider struct {
+	ctx           context.Context
+	authConfigs   v3.AuthConfigInterface
+	azureADClient *Client
+	userMGR       user.Manager
+}
+
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager) common.AuthProvider {
+	azureADClient := &Client{
+		httpClient: &http.Client{},
+	}
+	return &azureADProvider{
+		ctx:           ctx,
+		authConfigs:   mgmtCtx.Management.AuthConfigs(""),
+		azureADClient: azureADClient,
+		userMGR:       userMGR,
+	}
+}
+
+func (p *azureADProvider) GetName() string {
+	return Name
+}
+
+func (p *azureADProvider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = p.actionHandler
+	schema.Formatter = p.formatter
+}
+
+func (p *azureADProvider) TransformToAuthProvider(authConfig map[string]interface{}) map[string]interface{} {
+	azureADP := common.TransformToAuthProvider(authConfig)
+	return azureADP
+}
+
+func (p *azureADProvider) AuthenticateUser(input interface{}) (v3.Principal, []v3.Principal, map[string]string, error) {
+	azureADCredential, ok := input.(*v3public.BasicLogin)
+	if !ok {
+		return v3.Principal{}, nil, nil, errors.New("unexpected input type")
+	}
+
+	return p.loginUser(azureADCredential, nil, false)
+}
+
+func (p *azureADProvider) SearchPrincipals(searchkey, principalType string, token v3.Token) ([]v3.Principal, error) {
+	var principals []v3.Principal
+	var err error
+
+	config, err := p.getAzureADConfig()
+	if err != nil {
+		return principals, nil
+	}
+	azureAccessToken := token.ProviderInfo["access_token"]
+
+	var useraccts []Account
+	var groupaccts []Account
+	if principalType == "" || principalType == "user" {
+		principalType = "user"
+		users, err := p.azureADClient.searchUsers(searchkey, azureAccessToken, config)
+		if err != nil {
+			return nil, err
+		}
+		useraccts = append(useraccts, users...)
+		for _, acct := range useraccts {
+			p := p.toPrincipal(principalType, acct, nil)
+			principals = append(principals, p)
+		}
+	}
+	if principalType == "" || principalType == "group" {
+		principalType = "group"
+		groups, err := p.azureADClient.searchGroups(searchkey, azureAccessToken, config)
+		if err != nil {
+			return nil, err
+		}
+		groupaccts = append(groupaccts, groups...)
+		for _, acct := range groupaccts {
+			p := p.toPrincipal(principalType, acct, nil)
+			principals = append(principals, p)
+		}
+	}
+
+	for _, principal := range principals {
+		if principal.PrincipalType == "user" {
```

Comment:
maybe we can add constants for user/group scope like github/ad do?

Reply:
ok